var/spack/repos/builtin/packages/elpa/package.py

```diff
@@ -62,21 +65,48 @@ def libs(self):
 
     build_directory = 'spack-build'
 
-    def setup_environment(self, spack_env, run_env):
+    def configure_args(self):
```

Comment:
In which sense did you "fix the use" of `setup_environment`? If you can explain clearly where the process was failing, that may help expose bugs, but honestly this is not clear at a first glance to me. Also note that:
```
root@module-file-tutorial:/# spack install elpa ^openblas ^mpich %gcc@7.2.0
==> netlib-scalapack is already installed in /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-7.2.0/netlib-scalapack-2.0.2-c4v5l7jpkphoxbg34bviebl6cbkm5e6h
==> mpich is already installed in /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-7.2.0/mpich-3.2-7gxffhvkjslwjtvxnnofdx6m53mxhkg5
==> openblas is already installed in /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-7.2.0/openblas-0.2.20-kvddidenzbi6calqyxg6lxt65vdi5lw7
==> Installing elpa
==> Warning: Suspicious requests to set or unset 'CC' found
==> Warning: 	    	env.set('CC', join_path(link_dir, compiler.link_paths['cc'])) at /usr/local/lib/spack/spack/build_environment.py:145
==> Warning: 	--->	spack_env.set('CC', spec['mpi'].mpicc) at /usr/local/var/spack/repos/builtin/packages/elpa/package.py:70
==> Warning: Suspicious requests to set or unset 'CXX' found
==> Warning: 	    	env.set('CXX', join_path(link_dir, compiler.link_paths['cxx'])) at /usr/local/lib/spack/spack/build_environment.py:148
==> Warning: 	--->	spack_env.set('CXX', spec['mpi'].mpicxx) at /usr/local/var/spack/repos/builtin/packages/elpa/package.py:72
==> Warning: Suspicious requests to set or unset 'FC' found
==> Warning: 	    	env.set('FC', join_path(link_dir, compiler.link_paths['fc'])) at /usr/local/lib/spack/spack/build_environment.py:154
==> Warning: 	--->	spack_env.set('FC', spec['mpi'].mpifc) at /usr/local/var/spack/repos/builtin/packages/elpa/package.py:71
==> Fetching http://elpa.mpcdf.mpg.de/html/Releases/2016.05.004/elpa-2016.05.004.tar.gz
######################################################################## 100.0%
==> Staging archive: /usr/local/var/spack/stage/elpa-2016.05.004-nswub3qbro2oe436iyelp2ijct7varpv/elpa-2016.05.004.tar.gz
==> Created stage in /usr/local/var/spack/stage/elpa-2016.05.004-nswub3qbro2oe436iyelp2ijct7varpv
==> No patches needed for elpa
==> Building elpa [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed elpa
  Fetch: 0.37s.  Build: 43.08s.  Total: 43.46s.
[+] /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-7.2.0/elpa-2016.05.004-nswub3qbro2oe436iyelp2ijct7varpv
```
So the issues are less generic than `%gcc@7.2.0` (as you claim in the description).

Reply:
The commit message was not chosen very well and I was trying to debug the problems described in #6143 and tentatively attributed the issue to the use of `setup_environment`. I created the PR following the suggestion [here](https://github.com/spack/spack/pull/5306/files#r149293752) without too much cleaning up first. I apologize.

Like what you showed with a different set of dependencies, the issue I was encountering may not always appear. Please take a look at the linked issue #6143 & #6144. I have detailed log output there. It feels very strange that the second time `setup_environment` was called the spec object no longer have a mpi spec.

Reply:
> I apologize.

Hey, no need to apologize! :smile: 

The point I am trying to make is that probably there's a bug somewhere that affects how the environment is set, but there's definitely no misconception on the use of `setup_environment`. 

Imo, the fix here isn't a real fix, because it just hacks its way to prevent the symptoms to show up. But the bug is still there waiting for the next package. From my point of view it would be better to understand why things are not behaving as expected, and make the package work for you without the need to change `setup_environment`. I'll have a look at #6143 and #6144 asap.